### PR TITLE
Change limit from 5 to 10: Fixes #783

### LINKED ISSUE #783: When there are more than 5 lines, line 6 an up are invisible
### Steps to reproduce

Download and open this [HTML document](https://gist.github.com/openjck/0d4b779240975c1b1bd1f53d52b0a0ce)

### Actual result

The final 5 lines (in this case, the top-most 5 lines) are invisible. Hovering over them does cause a black circle to appear, however.

### Expected result

All lines are colored. d3 has [something](http://bl.ocks.org/aaizemberg/78bd3dade9593896a59d) that looks like it could help with this. As a last resort, if there are any lines that can't be assigned a color, having multiple black lines would be better than having insible lines.

### PR DESCRIPTION
**Pull request by shikhar-scs**
As discussed at #783, I am filing this request. It is incomplete though and @wlach has confirmed to assist me here.

### CODE REVIEW

File: dist/metricsgraphics.css [hunk 1]
```diff
@@ -226,6 +246,50 @@ tspan.hist-symbol {
     fill: #5c5c5c;
 }
```
Comment (wlach) [inline dist/metricsgraphics.css]: I think we should define a color called `.mg-line-color` here (repeating pattern for the other types) with a stroke color of black. We could then give the line (or whatever other object you're manipulating) that class and it should appear in that color as long as there isn't class overriding it (which it would be default for up to 10 lines, and possibly further on that if the user creates a user-defined style).

File: dist/metricsgraphics.css [hunk 2]
```diff
@@ -226,6 +246,50 @@ tspan.hist-symbol {
...
+.mg-hover-line6-color {
+    fill: steelblue;
+}
+
+.mg-line7-color {
+    stroke: #f673bf;;
+}
+
+.mg-hover-line7-color {
+    fill: #f673bf;;
```
[the remark this answers is not in the document]
Comment (shikhar-scs) [reply dist/metricsgraphics.css]: Thanx for pointing this out, I'll correct all of them.

File: dist/metricsgraphics.js [hunk 3]
```diff
@@ -6175,7 +6175,7 @@ function mg_color_point_mouseover(args, elem, d) {
     if (args.legend_target) {
 
       var div = d3.select(args.legend_target).append('div').classed('mg-bar-target-legend', true);
-      
+
```
[the remark this answers is not in the document]
Comment (shikhar-scs) [reply dist/metricsgraphics.js]: Yeah sure :+1:

File: dist/metricsgraphics.css (hunk 1 above)
Comment (shikhar-scs) [reply dist/metricsgraphics.css]: ok, I'll try and implement the same :+1:

**Comment by shikhar-scs** [comment]
@wlach I've made all the required changes and they do reflect now for `lines`. However, I am a bit unaware as to how shall I generate examples, it looks a bit difficult.
I had one for lines already and that had helped.

Here it can be noticed that after 10, the lines are black in color.

metrics-graphics

For legends I guess the changes need to be made here

https://github.com/mozilla/metrics-graphics/blob/296c39fb53ea1a941eacc5bcdc8bbcc917b13ec5/src/js/charts/line.js#L227-L235

But again, I cant see the changes as I dont know how to generate an example.

Similarly, I'll find where to make changes for other figures

**Comment by wlach** [comment]
@shikhar-scs do you mean an example of a legend?

File: dist/metricsgraphics.js [hunk 4]
```diff
@@ -4659,29 +4659,36 @@ MG.button_layout = function(target) {
...
+          this_path.classed('mg-line' + (line_id) + '-color', true);
+
+      else
+          this_path.classed('mg-line' + '-color', true);
   }
 
   function mg_color_line(args, this_path, which_line, line_id) {
-    if (args.colors) {
+      console.log(args.length);
```
Comment (shikhar-scs) [inline dist/metricsgraphics.js]: ignore this

File: dist/metricsgraphics.js [hunk 5]
```diff
@@ -4659,29 +4659,36 @@ MG.button_layout = function(target) {
...
       } else {
         // this_path.classed('mg-line' + (line_id) + '-color', true);
-        mg_default_color_for_path(this_path, line_id);
+        mg_default_color_for_path(args, this_path, line_id);
       }
     } else {
       // this is the typical workflow
       // this_path.classed('mg-line' + (line_id) + '-color', true);
-      mg_default_color_for_path(this_path, line_id);
+      mg_default_color_for_path(args, this_path, line_id);
```
Comment (shikhar-scs) [inline dist/metricsgraphics.js]: ignore this and the above 2 too

**Comment by shikhar-scs** [comment]
> @shikhar-scs do you mean an example of a legend?

Yeah, I mean in general how do I create an example, or if there are some handy examples to use

**Comment by wlach** [comment]
Good question, at the moment I have mostly been modifying the source to the version of metricsgraphics in dist/, creating some handcrafted html that includes that, and then copying the changes over at the end of that process.

This is obviously a pretty embarassing workflow-- let me see if I can't come up with something better today.

**Comment by shikhar-scs** [comment]
Yeah Sure, no problem. I'll wait.

P.S. are the changes in `line.js` fine?

**Comment by wlach** [comment]
@shikhar-scs ok I finally fixed up the `gulp serve` workflow which should be helpful for testing changes as you go. Have a look at https://github.com/mozilla/metrics-graphics#contributing :)

**Comment by shikhar-scs** [comment]
Thanks a lot @wlach for that, I'll get started with my work again :+1:

**Comment by shikhar-scs** [comment]
@wlach I've made the required changes in `lines.js`. Please have a look.

Ignore the indentation errors, I'll remove them once this is ready for merge.

File: dist/metricsgraphics.css [hunk 6]
```diff
@@ -5,6 +5,10 @@
     opacity: 0.8;
 }
 
+.mg-area-color {
+    fill: #000000;
```
Comment (wlach) [inline dist/metricsgraphics.css]: You should be able to get away with just `#000` here (and below).

File: dist/metricsgraphics.js [hunk 7]
```diff
@@ -4660,7 +4660,12 @@ MG.button_layout = function(target) {
   }
 
   function mg_default_color_for_path(this_path, line_id) {
-    this_path.classed('mg-line' + (line_id) + '-color', true);
+      if(line_id <= 10)
+
+          this_path.classed('mg-line' + (line_id) + '-color', true);
+
+      else
+          this_path.classed('mg-line' + '-color', true);
```
Comment (wlach) [inline dist/metricsgraphics.js]: With this approach, lines of a greater number than 10 will be always be black, which isn't quite what we want. I think the right approach here is to give each line both the mg-line-color AND the mg-line-X-color. That gives the user the opportunity to extend the set of supported colors while still defaulting to black when that isn't present.

Comment (shikhar-scs) [reply dist/metricsgraphics.js]: Yeah you are right. I'll implement the same.

**Comment by shikhar-scs** [comment]
@wlach I've made all required changes to `src/js/charts/line.js` and `dist/metricsgraphics.js`. Please have a look at it.

File: dist/metricsgraphics.css (hunk 6 above)
Comment (shikhar-scs) [reply dist/metricsgraphics.css]: Done

File: src/js/charts/line.js [hunk 8]
```diff
@@ -231,7 +243,8 @@
             legend_text.attr('fill', args.colors[which_line]);
           }
         } else {
-          legend_text.classed('mg-line' + (line_id) + '-legend-color', true);
+              legend_text.classed('mg-line' + (line_id) + '-legend-color', true);
+              legend_text.classed('mg-line' + '-legend-color', true);
         }
 
         mg_prevent_vertical_overlap(plot.legend_group.selectAll('.mg-line-legend text').nodes(), args);
```
Comment (shikhar-scs) [inline src/js/charts/line.js]: Remember to look at this file too as it isnt displayed by default because of high number of changes

**Comment by shikhar-scs** [comment]
@wlach mind reviewing

**Comment by wlach** [comment]
Yep, I will take a look tomorrow.

File: src/js/charts/line.js [hunk 9]
```diff
@@ -1,153 +1,165 @@
 (function() {
-  'use strict';
+    'use strict';
```
Comment (wlach) [inline src/js/charts/line.js]: Please use the 2-space indentation that's the mg default, both here and below.

**Comment by shikhar-scs** [comment]
> I don't mean to be a pain, but I'd really appreciate it if you would remove any change which isn't directly related to the modification to functionality we're talking about in this bug -- this needs to be done before we land in any case, so it won't be effort wasted.

Yep Sure

@wlach I've made the best possible `files` now and also squashed my commits. please preview :+1:

File: dist/metricsgraphics.css [hunk 10]
```diff
@@ -257,6 +334,31 @@ tspan.hist-symbol {
...
+.mg-line9-legend-color {
+    color: #006400;
+    fill: #006400;
+}
+
+.mg-line10-legend-color {
+    color: #92514f;
+    fill: #92514f;
+}
+
```
Comment (shikhar-scs) [inline dist/metricsgraphics.css]: all the css has been included on the basis of logic only and is pretty easy to understand :smile:

File: dist/metricsgraphics.js [hunk 11]
```diff
@@ -4540,6 +4540,8 @@ MG.button_layout = function(target) {
   function mg_line_color_text(elem, d, args) {
     elem.classed('mg-hover-line' + d.line_id + '-color', args.colors === null)
       .attr('fill', args.colors === null ? '' : args.colors[d.line_id - 1]);
+    elem.classed('mg-hover-line' + '-color', args.colors === null)
+      .attr('fill', args.colors === null ? 'black' : args.colors[d.line_id - 1]);
   }
```
Comment (shikhar-scs) [inline dist/metricsgraphics.js]: logic for `mg-hover-line`

File: dist/metricsgraphics.js [hunk 12]
```diff
@@ -4653,6 +4655,14 @@ MG.button_layout = function(target) {
...
           .attr('clip-path', 'url(#mg-plot-window-' + mg_target_ref(args.target) + ')');
+
+        svg.append('path')
+          .classed('mg-main-area', true)
+          .classed('mg-area' + line_id, true)
+          .classed('mg-area' + '-color', args.colors === null)
+          .attr('d', plot.area(args.data[which_line]))
+          .attr('fill', args.colors === null ? '' : args.colors[line_id - 1])
+          .attr('clip-path', 'url(#mg-plot-window-' + mg_target_ref(args.target) + ')');
       }
```
Comment (shikhar-scs) [inline dist/metricsgraphics.js]: logic for `mg-area`

File: dist/metricsgraphics.js [hunk 13]
```diff
@@ -4661,6 +4671,7 @@ MG.button_layout = function(target) {
 
   function mg_default_color_for_path(this_path, line_id) {
     this_path.classed('mg-line' + (line_id) + '-color', true);
+    this_path.classed('mg-line' + '-color', true);
   }
```
Comment (shikhar-scs) [inline dist/metricsgraphics.js]: logic for `mg-line-color`

File: dist/metricsgraphics.js [hunk 14]
```diff
@@ -4768,6 +4779,7 @@ MG.button_layout = function(target) {
           }
         } else {
           legend_text.classed('mg-line' + (line_id) + '-legend-color', true);
+          legend_text.classed('mg-line' + '-legend-color', true);
         }
```
Comment (shikhar-scs) [inline dist/metricsgraphics.js]: logic for `mg-line-legend-color`

File: examples/js/lib/ace.js [hunk 15]
```diff
@@ -1,11 +0,0 @@
...,s),this.isEmpty=function(){return this.$isEmpty||this.anchor.row==this.lead.row&&this.anchor.column==this.lead.column},this.isMultiLine=function(){return this.isEmpty()?!1:this.getRange().isMultiLine()},this.getCursor=function(){return this.lead.getPosition()},this.setSelectionAnchor=function(e,t){this.anchor.setPosition(e,t),this.$isEmpty&&(this.$isEmpty=!1,this._emit("changeSelection"))},this.getSelectionAnchor=function(){return this.$isEmpty?this.getSelectionLead():this.anchor.getPosition()},this.getSelectionLead=function(){return this.lead.getPosition()},this.shiftSelection=function(e){if(this.$isEmpty){this.moveCursorTo(this.lead.row,this.lead.column+e);return}var t=this.getSelectionAnchor(),n=this.getSelectionLead(),r=this.isBackwards();(!r||t.column!==0)&&this.setSelectionAnchor(t.row,t.column+e),(r||n.column!==0)&&this.$moveSelection(function(){this.moveCursorTo(n.row,n.column+e)})},this.isBackwards=function(){var e=this.anchor,t=this.lead;return e.row>t.row||e.row==t.row&&e.column>t.column},this.getRange=function(){var e=this.anchor,t=this.lead;return this.isEmpty()?o.fromPoints(t,t):this.isBackwards()?o.fromPoints(t,e):o.fromPoints(e,t)},this.clearSelection=function(){th
```
Comment (shikhar-scs) [inline examples/js/lib/ace.js]: This file has magically reappeared while, squashing. and it had already been deleted initially :smile: so I guess this can be safely merged :+1:

File: src/js/charts/line.js [hunk 16]
```diff
@@ -641,7 +653,11 @@
...
 
-      d.values.forEach(function(datum) {
+      d.values.forEach(function(datum, index, list) {
+        if (args.missing_is_hidden && list[index]['_missing']) {
+          return;
+        }
+
         if (mg_data_in_plot_bounds(datum, args)) mg_update_aggregate_rollover_circle(args, svg, datum);
       });
     } else if ((args.missing_is_hidden && d['_missing']) || d[args.y_accessor] === null) {
```
Comment (shikhar-scs) [inline src/js/charts/line.js]: all respective changes as mentioned in `metrics-graphics.js` file above :smile:

File: src/js/charts/line.js [hunk 17]
```diff
@@ -4,6 +4,8 @@
   function mg_line_color_text(elem, d, args) {
     elem.classed('mg-hover-line' + d.line_id + '-color', args.colors === null)
       .attr('fill', args.colors === null ? '' : args.colors[d.line_id - 1]);
+        elem.classed('mg-hover-line' + '-color', args.colors === null)
+          .attr('fill', args.colors === null ? 'black' : args.colors[d.line_id - 1]);
```
Comment (wlach) [inline src/js/charts/line.js]: The logic to make this black or not should not be conditional on the content of `args`. We should always apply this style regardless. 

Logically, this statement should be before the custom style is applied. It should be indented 2 spaces, not 4.

Also there is no need to add two strings together when they are constant types. I.e.: `'mg-hover-line' + '-color'` -> `'mg-hover-line-color'`.

File: src/js/charts/line.js [hunk 18]
```diff
@@ -117,6 +119,14 @@
...
           .attr('fill', args.colors === null ? '' : args.colors[line_id - 1])
           .attr('clip-path', 'url(#mg-plot-window-' + mg_target_ref(args.target) + ')');
+
+        svg.append('path')
+          .classed('mg-main-area', true)
+          .classed('mg-area' + line_id, true)
+          .classed('mg-area' + '-color', args.colors === null)
+          .attr('d', plot.area(args.data[which_line]))
+          .attr('fill', args.colors === null ? '' : args.colors[line_id - 1])
+          .attr('clip-path', 'url(#mg-plot-window-' + mg_target_ref(args.target) + ')');
```
Comment (wlach) [inline src/js/charts/line.js]: Could you explain what this is supposed to be doing? It doesn't seem correct to me (again depending on args when we're trying to add a "default" style which doesn't depend on them).

File: src/js/charts/line.js [hunk 19]
```diff
@@ -125,6 +135,7 @@
 
   function mg_default_color_for_path(this_path, line_id) {
     this_path.classed('mg-line' + (line_id) + '-color', true);
+    this_path.classed('mg-line' + '-color', true);
```
Comment (wlach) [inline src/js/charts/line.js]: Similar comments as those above on `mg_line_color_text`  apply here.

File: src/js/charts/line.js [hunk 20]
```diff
@@ -232,6 +243,7 @@
           }
         } else {
           legend_text.classed('mg-line' + (line_id) + '-legend-color', true);
+          legend_text.classed('mg-line' + '-legend-color', true);
```
Comment (wlach) [inline src/js/charts/line.js]: `'mg-line' + '-legend-color'` -> `'mg-line-legend-color'`

File: src/js/charts/line.js [hunk 21]
```diff
@@ -641,7 +653,11 @@
       svg.selectAll('circle.mg-line-rollover-circle')
         .style('opacity', 0);
 
-      d.values.forEach(function(datum) {
+      d.values.forEach(function(datum, index, list) {
+        if (args.missing_is_hidden && list[index]['_missing']) {
+          return;
+        }
```
Comment (wlach) [inline src/js/charts/line.js]: Can you explain what this does? It doesn't see obvious to me.

File: dist/metricsgraphics.js [hunk 22]
```diff
@@ -6175,7 +6187,6 @@ function mg_color_point_mouseover(args, elem, d) {
     if (args.legend_target) {
 
       var div = d3.select(args.legend_target).append('div').classed('mg-bar-target-legend', true);
-
```
Comment (wlach) [inline dist/metricsgraphics.js]: Remove this change.

File: dist/metricsgraphics.js [hunk 23]
```diff
@@ -4540,6 +4540,8 @@ MG.button_layout = function(target) {
   function mg_line_color_text(elem, d, args) {
```
Comment (wlach) [inline dist/metricsgraphics.js]: Please revert the changes to this file.

File: examples/js/lib/ace.js (hunk 15 above)
Comment (wlach) [inline examples/js/lib/ace.js]: Please revert changes to this file (I removed it in another commit, maybe you just need to rebase on master?).

File: src/js/charts/line.js [hunk 24]
```diff
@@ -231,6 +243,7 @@
             legend_text.attr('fill', args.colors[which_line]);
           }
         } else {
+          legend_text.classed('mg-line-legend-color', true);
           legend_text.classed('mg-line' + (line_id) + '-legend-color', true);
         }
```
Comment (shikhar-scs) [inline src/js/charts/line.js]: @wlach Please have a look at this now

**Comment by wlach** [comment]
@shikhar-scs the code looks pretty reasonable (mostly), but travis now seems to be failing ^^ can you investigate? You should be able to run the tests locally via `gulp test`. I suspect it's one of the javascript changes. Thanks for your patience on this so far.

File: src/js/charts/line.js [hunk 25]
```diff
@@ -110,6 +112,14 @@
           .attr('d', plot.area(args.data[which_line]))
           .attr('clip-path', 'url(#mg-plot-window-' + mg_target_ref(args.target) + ')');
       } else { // otherwise, add the area
+        svg.append('path')
+          .classed('mg-main-area', true)
+          .classed('mg-area' + line_id, true)
+          .classed('mg-area-color')
+          .attr('d', plot.area(args.data[which_line]))
+          .attr('fill', 'black')
+          .attr('clip-path', 'url(#mg-plot-window-' + mg_target_ref(args.target) + ')');
```
Comment (wlach) [inline src/js/charts/line.js]: This is the one part of the code that I'm unsure about (I think it might be what's causing the unit tests to fail). It doesn't seem correct to basically copypaste this rather large block of code twice. If you need to add a class to the path, can you not do that by appending a classed statement to the original set of calls below?

File: src/js/charts/line.js [hunk 26]
```diff
@@ -124,6 +134,7 @@
   }
 
   function mg_default_color_for_path(this_path, line_id) {
+    this_path.classed('mg-line-color', true);
     this_path.classed('mg-line' + (line_id) + '-color', true);
```
Comment (wlach) [inline src/js/charts/line.js]: Now that I look at it, could you not write this as:

```js
this_path.classed('mg-line-color', true)
              .classed('mg-line' + (line_id) + '-color', true);
```

?

Comment (shikhar-scs) [reply src/js/charts/line.js]: yeah that looks better, I'll put it that way :+1:

File: src/js/charts/line.js (hunk 25 above)
Comment (shikhar-scs) [reply src/js/charts/line.js]: Yeah that's exact place where the problem was occurring. Removing that code passed all the tests.
I'll try and figure out a way to correct that.

Comment (shikhar-scs) [reply src/js/charts/line.js]: Yep, 100% sure that `fill black ` is causing the error. What can be substituted in that place ?

Comment (wlach) [reply src/js/charts/line.js]: Offhand, I am not sure and don't have time to test right now. Maybe it's just the tests that need to be updated? I would experiment a bit. If you're still stuck after trying some stuff let me know and I'll see if I can help.

Comment (shikhar-scs) [reply src/js/charts/line.js]: The problem has been solved though !

The classes being pushed in this 
https://github.com/shikhar-scs/metrics-graphics/blob/c063971799a5d2eaaf35c7ad7660ba793a9abe5e/src/js/charts/line.js#L4-L9

and

https://github.com/shikhar-scs/metrics-graphics/blob/c063971799a5d2eaaf35c7ad7660ba793a9abe5e/src/js/charts/line.js#L115-L121


already have the `fill` attribute set to black , i.e. look at `mg-hover-line-color` and `mg-area-color`. So, basically we dont need to overwrite them further . 


I'm pushing the final (hopefully) code. Please review :+1:

**Comment by shikhar-scs** [comment]
The builds pass please review

File: src/js/charts/line.js [hunk 27]
```diff
@@ -2,6 +2,7 @@
   'use strict';
 
   function mg_line_color_text(elem, d, args) {
+    elem.classed('mg-hover-line-color')
     elem.classed('mg-hover-line' + d.line_id + '-color', args.colors === null)
```
Comment (wlach) [inline src/js/charts/line.js]: Could you do the same trick as below when calling .classed?

File: src/js/charts/line.js [hunk 28]
```diff
@@ -231,6 +234,7 @@
             legend_text.attr('fill', args.colors[which_line]);
           }
         } else {
+          legend_text.classed('mg-line-legend-color', true);
           legend_text.classed('mg-line' + (line_id) + '-legend-color', true);
```
Comment (wlach) [inline src/js/charts/line.js]: Same as above.

**Comment by shikhar-scs** [comment]
done

**Comment by shikhar-scs** [comment]
@wlach ?

**Comment by wlach** [comment]
@shikhar-scs sorry about the late reply again, I somehow missed the fixes. looks good except for the lack of rebasing. if you have time to do that today, please do. otherwise I can do it myself tomorrow before merging.

**Comment by shikhar-scs** [comment]
Would be good if you do that while rebasing :smile:

**Comment by wlach** [comment]
Merged manually, thank you! I also updated the README in 36c2f4c8a51aea660856261e4540e0eac6a21d7c

**Comment by shikhar-scs** [comment]
Thankyou @wlach

**Comment by shikhar-scs** [comment]
@wlach why does this PR show its status as `Closed with unmerged commits` . Was it not merged in the usual way ? Also, in the README.md file the color `steelblue` has been mentioned to be used in case of a new line. `Steelblue` has already been used in case of the `6th line`. So lets mention some new color there.

**Comment by wlach** [comment]
@shikhar-scs I merged it manually so I could rebase it on master and update the commit message. If you look at the changelog you'll see your commit there:

https://github.com/mozilla/metrics-graphics/commit/fcbc2fdcc81beb5633b547c7a8a81de2f0beb4b3

For the README issue please file a followup PR!

**Comment by shikhar-scs** [comment]
Yeah I see that 😅. Thanks again .

> For the README issue please file a followup PR!

Yeah sure I'll do that soon
